Update CMSIS to 5.4.0

c1728p9 opened the pull request:
### Description

Pull in the newest version of CMSIS - 5.4.0.

In specific this patch does the following:
- Preparation for importing
    - Remove use of uVisor hooks
    - Move Cortex A customization out for RTX and into mbed
    - Update importer docs and cmsis_importer.json in prep for update
- Run importer.py to pull in the updated version of RTX/CMSIS_5
- Add additional RTX fixes
    - ~CMSIS/RTX: Conditionally enable TZ~ CMSIS/RTX: Patch to conditionally compile
    - CMSIS/RTX: Remove the file os_tick_gtim.c
    - Fix define around EvrRtxKernelInitialized ([Upstreamed into CMSIS_5](https://github.com/ARM-software/CMSIS_5/pull/410))

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Breaking change
    [x] CMSIS update

bulislaw committed:
CMSIS/RTX: Patch RTX4 to preserve osThreadDef compatibility

mbed OS used older RTX4 version and with osThreadDef accepting only 3
parameters, to preserve compatibility we hardcode the 'instances'
parameter to 1.

(cherry picked from commit 428acae1b2ac15c3ad523e8d40755a9301220822)
(cherry picked from commit 4360b7bbf815c4d812005938c9c27af199803a97)

bulislaw committed:
CMSIS/RTX: Patch RTX so irq_cm4f.s files work with no FPU targets

(cherry picked from commit cc2e0517e1c6440abf88f2815b8e1501a55cdd4d)

bulislaw committed:
CMSIS/RTX: Allow overwriting mutex ops for ARMC

(cherry picked from commit b88254809eb626689c8aeb41304a308bf4e34a04)

Commit:
CMSIS/RTX: Pre-processor defines used for assembly

CMSIS repo does not support pre-processor defines, hence multiple assembly
files are added for secure/non-secure and floating point tools.

Mbed OS tools support assembly file pre-processing, but the build system
does not support multiple assembly files for each target, hence updating
the assembly files.

(cherry picked from commit 287121ffdc4c9c19f9ce8872e4edd941862daca1)

Commit:
CMSIS/RTX: Patch to conditionally compile

tz_context.c should be compiled only for secure world,
definition of API's in tz_context.h should be part of secure
binary/bootloader when building mbed-os as non-secure

(Cherry picked from d0a43b8af0eef4775ec4c3da5994ecceb9ed4558)

bulislaw committed:
CMSIS/RTX: Fix using FALSE/TRUE with preprocesor

(cherry picked from commit 1752803626865147dca92f30a39cef8d04581736)

bulislaw committed:
CMSIS/RTX: Reintroduce arm_math.h

(cherry picked from commit 6a6e3ac0ebab1a6b6aa08d0928702c79562acee9)

c1728p9 committed:
Fix define around EvrRtxKernelInitialized

Rename the define EVR_RTX_KERNEL_INITIALIZE_COMPLETED_DISABLE to
EVR_RTX_KERNEL_INITIALIZED_DISABLE so it matches the rest of RTX.

c1728p9 committed:
Move Code to get Cortex A tick irqn

Move the function "mbed_get_a9_tick_irqn()" declared in os_tick.h
to SysTimer.h so it does not get overridden when RTX is updated.

This function was originally added in:
3d3e89097d5db0e24075134638d81c5f859c96c1
Fixes for RZ_A1H issue 6543

c1728p9 committed:
CMSIS/RTX: Remove the file os_tick_gtim.c

Cortex-A devices in mbed-os supply their own OS timer so the code
from RTX is not needed.

c1728p9 committed:
Revert uVisor changes for K64F

Revert the changes made in the commit:
1c2a7d8842696a4c772ceb1437e08b8adb0689c8
Added the commit: Access MCG and SIM through secure access

c1728p9 committed:
Revert "EFM32: Use SECURE_ACCESS to access the ROM table"

This reverts commit 1b0922e97bb9f9147727c4b578c037d34da2d0a4.

c1728p9 committed:
Update cmsis_importer in prep for CMSIS 5.4.0

Reference the correct commits to pull in and remove commits related to
uVisor since these are no longer needed. Also add the missing file
cmsis_os1.c to cmsis_importer.json.

Below is detailed information on each commit added to cmsis_importer.json

CMSIS/RTX: Patch RTX4 to preserve osThreadDef compatibility
4360b7bbf815c4d812005938c9c27af199803a97

CMSIS/RTX: Patch RTX so irq_cm4f.s files work with no FPU targets
cc2e0517e1c6440abf88f2815b8e1501a55cdd4d

CMSIS/RTX: Allow overwriting mutex ops for ARMC
b88254809eb626689c8aeb41304a308bf4e34a04

CMSIS/RTX: Pre-processor defines used for assembly
287121ffdc4c9c19f9ce8872e4edd941862daca1

CMSIS/RTX: Fix using FALSE/TRUE with preprocesor
1752803626865147dca92f30a39cef8d04581736

CMSIS/RTX: Reintroduce arm_math.h
6a6e3ac0ebab1a6b6aa08d0928702c79562acee9

fixup

c1728p9 committed:
Add details to importer script docs

Add an example showing how to run the importer and a note to indicate
that it must be run from the mbed-os directory.

cmonr commented:
Testing.

/morph build

0xc0170 commented:
There are linker failures (does this change any API? I found at least one there: signalSet/wait in the CI logs).

donatieng reviewed (changes requested):
Looks like TCP/UDP tests use CMSIS-OS v1 APIs which are implemented in `rtos/TARGET_CORTEX/rtx4/cmsis_os1.c` which has been removed. Otherwise looks good :).

c1728p9 committed:
[CMSIS_5]: Updated to 0b521765

c1728p9 commented:
I made the following changes:
- updated cmsis_importer.json to include cmsis_os1.c
- re-ran importer.py to pull in the updated version of RTX/CMSIS_5

deepikabhavnani commented:
@ARMmbed/mbed-os-maintainers - Please note this PR will remove uVisor changes done on top of CMSIS, hence and should be merged after https://github.com/ARMmbed/mbed-os/pull/7592

cmonr commented:
@donatieng Mind taking an updated look?

Running CI to check if PR has improved.
/morph build

c1728p9 committed:
Fix CMSIS __RESTRICT define for IAR

Revert the define __RESTRICT from "__restrict" to "restrict". This
fixes compilation for IAR.

c1728p9 commented:
Updated this PR to fix IAR compilation.

cmonr commented:
/morph build

cmonr commented:
Single failure, but unable to determine _why_...
/morph build

c1728p9 commented:
Replaced the commit "CMSIS/RTX: Conditionally enable TZ" with @deepikabhavnani 's
commit "CMSIS/RTX: Patch to conditionally compile" to fix TZ builds
on the NUMAKER_PFM_M2351.

cmonr commented:
/morph build

cmonr commented:
@c1728p9 You might need to take a look at the travis tests.

cmonr commented:
https://github.com/ARMmbed/mbed-os/pull/7592 has now been merged. This will need a rebase.

c1728p9 commented:
Rebased and resolved conflicts

c1728p9 committed:
Remove double free RTOS tests

Remove the double free RTOS tests since this was never defined
behavior of CMSIS-RTOS. This allows testing to pass.

The RTX commit which caused this test to start failing is:
c3b123ef4256f65537e2597af475fc20ec9a383e
RTX5: updated MemoryPoolFree (removed count check) [SDCMSIS-801]

Note - Double freeing an element from a memory pool was never safe.
The error return value when double freeing was misleading since memory
corruption may still be occurring in that case. For more information
on this see SDCMSIS-801.

c1728p9 commented:
Remove the double free RTOS tests since this was never defined behavior of CMSIS-RTOS. This allows testing to pass.

The RTX commit which caused this test to start failing is:
https://github.com/ARM-software/CMSIS_5/commit/c3b123ef4256f65537e2597af475fc20ec9a383e
`RTX5: updated MemoryPoolFree (removed count check) [SDCMSIS-801]`

Note - Double freeing an element from a memory pool was never safe. The error return value when double freeing was misleading since memory corruption may still be occurring in that case. For more information on this see SDCMSIS-801.

CC @RobertRostohar

c1728p9 commented:
/morph build

cmonr commented:
The prep step got stuck somehow.

/morph build

cmonr commented:
Unclear if test failure was caused by PR or not.

/morph test

cmonr commented:
@c1728p9 Would you be able to take a look at @donatieng comment? Is the file still not in the PR?

c1728p9 commented:
Hi @cmonr I added `cmsis_os1.c` back to this PR earlier, so this should be good to go.

cmonr commented:
@donatieng Mind re-reviewing? This PR should be good to go otherwise.

cmonr commented:
@donatieng Mind re-reviewing/OKing?

donatieng reviewed (approved):
Looks good! Thanks @c1728p9
